File: BchtSystemSpeed/app/src/main/cpp/camera/RaderCmd.cpp
//
// Created by admin on 2017/6/7.
//

#include <termios.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <jni.h>
#include <errno.h>
#include<grp.h>
#include<stdio.h>
#include <vector>
#include <string>
#include <dirent.h>
#include <sys/select.h>

#include "RaderCmd.h"
#include <stdio.h>






const unsigned short Radar::RadarCmd::Null = 0x0;
const unsigned short Radar::RadarCmd::STATUS_SEARCH = 0x8310;
const unsigned short Radar::RadarCmd::RESET = 0x8320;
const unsigned short Radar::RadarCmd::STATUS_SET = 0x8350;
const unsigned short Radar::RadarCmd::SOFTWARE_VERSION_SEARCH = 0x8351;
const unsigned short Radar::RadarCmd::PRODUCT_NO_SEARCH = 0x8352;
const unsigned short Radar::RadarCmd::PRAMETER_SET = 0x8355;
const unsigned short Radar::RadarCmd::PRAMETER_GET = 0x8356;
const unsigned short Radar::RadarCmd::PRAMETER_SAVE = 0x8357;
const unsigned short Radar::RadarCmd::RESTORE = 0x8359;
const unsigned short Radar::RadarCmd::SPEED   = 0x8A0A;
const unsigned short Radar::RadarCmd::STATUS_RESPONSE_SUCCESS   = 0x853A;
const unsigned short Radar::RadarCmd::STATUS_RESPONSE_FAILURE   = 0x85FF;
const unsigned short Radar::RadarCmd::MEASUREMENT_MODE =0x8410;
const unsigned short Radar::RadarCmd::TEST_MODE =0x8420;

const unsigned short Radar::StatusSearchResponse::STATUS_PARAMETER_SET     = 0x8801;
const unsigned short Radar::StatusSearchResponse::STATUS_SPEED_TEST_SET    = 0x8820;

const unsigned char Radar::StatusSetCmd::MODE_SET     = 0x01;
const unsigned char Radar::StatusSetCmd::MODE_SPEED_TEST    = 0x02;

const unsigned char Radar::MeasureSetCmd::METER_MODE     = 0x01;
const unsigned char Radar::MeasureSetCmd::EXIT_MODE    = 0x02;

const unsigned char Radar::TestMeasureSetCmd::DIREC_MODE     = 0x01;
const unsigned char Radar::TestMeasureSetCmd::SIMULATE_MODE    = 0x02;
const unsigned char Radar::TestMeasureSetCmd::MEATURE_MODE    = 0x03;

Radar::Header::Header()
        : header1(0xFF)
        , header2(0xFA)
        , dataLength(0)
        , cmd(RadarCmd::Null)
{

}

unsigned char Radar::Header::getHerder1() const {
    return header1;
}

void Radar::Header::setHerder1(unsigned char herder1) {
    this->header1 = herder1;
}

unsigned char Radar::Header::getHerder2() const {
    return header2;
}

void Radar::Header::setHerder2(unsigned char herder2) {
    this->header2 = herder2;
}

unsigned char Radar::Header::getDataLength() const {
    return dataLength;
}

void Radar::Header::setDataLength(unsigned char dataLength) {
    this->dataLength = dataLength;
}

unsigned short Radar::Header::getCmd() const {
    return cmd;
}

void Radar::Header::setCmd(unsigned short cmd) {
    this->cmd = cmd;
}

int Radar::Header::isValidHeader1() const
{
    return this->header1 == 0xFF ? 0 : ERROR_HEADER1;
}

int Radar::Header::isValidHeader2() const
{
    return this->header2 == 0xFA ? 0 : ERROR_HEADER2;
}

int Radar::Header::isValidLength() const
{
    return this->dataLength > 0 ? 0 : ERROR_DATALENGTH;
}

int Radar::Header::isValidCmd(unsigned short cmd) const
{
    return this->cmd == cmd ? 0 : ERROR_CMD;
}

Radar::StatusSearchCmd::StatusSearchCmd()
        : reserved(0x0)
{
    setCmd(RadarCmd::STATUS_SEARCH);
    setDataLength(0x03);
    checksum = Header::getChecksum(getCmd()) + Header::getChecksum(getDataLength()) + Header::getChecksum(reserved);
}

unsigned char Radar::StatusSearchCmd::getChecksum() const {
    return checksum;
}


Radar::StatusSearchResponse::StatusSearchResponse()
        : status(0)
        , checksum(0)
{

}

unsigned short Radar::StatusSearchResponse::getStatus() const {
    return status;
}

unsigned char Radar::StatusSearchResponse::getChecksum() const {
    return checksum;
}

int Radar::StatusSearchResponse::isValidStatus() const
{
    return (STATUS_PARAMETER_SET == status || STATUS_SPEED_TEST_SET == status) ? 0 : ERROR_RESPONSE_STATUS;
}

int Radar::StatusSearchResponse::isValidChecksum() const
{
    unsigned char validChecksum = Header::getChecksum(getCmd()) + Header::getChecksum(getDataLength()) + Header::getChecksum(status);
    return checksum == validChecksum ? 0 : ERROR_CHECKSUM;
}

bool Radar::StatusSearchResponse::isSuccess() const
{
    return 0 == isValidStatus();
}



Radar::ResetCmd::ResetCmd()
        : checksum(0)
{
    setCmd(RadarCmd::RESET);
    setDataLength(0x02);
    getChecksum();
}

unsigned char Radar::ResetCmd::getChecksum() {
    return checksum = Header::getChecksum(getCmd()) + Header::getChecksum(getDataLength());
}

int Radar::ResetCmd::isValidChecksum() const
{
    unsigned char validChecksum = Header::getChecksum(getCmd()) + Header::getChecksum(getDataLength());
    return checksum == (unsigned char)validChecksum ? 0 : ERROR_CHECKSUM;
}

Radar::Response::Response()
        : status(0)
        , checksum(0)
{}

unsigned short Radar::Response::getStatus() const {
    return status;
}

unsigned char Radar::Response::getChecksum() const {
    return checksum;
}

int Radar::Response::isValidStatus() const
{
    return (this->status == RadarCmd::STATUS_RESPONSE_SUCCESS || this->status == RadarCmd::STATUS_RESPONSE_FAILURE ) ? 0 : ERROR_RESPONSE_STATUS;
}

bool Radar::Response::isSuccess() const
{
    return this->status == RadarCmd::STATUS_RESPONSE_SUCCESS;
}

int Radar::Response::isValidChecksum() const
{
    unsigned char validChecksum = Header::getChecksum(getDataLength()) + Header::getChecksum(getCmd()) +  Header::getChecksum(status);
    return checksum == validChecksum ? 0 : ERROR_CHECKSUM;
}


Radar::StatusSetCmd::StatusSetCmd()
        : mode(0)
        , checksum(0)
{
    setCmd(RadarCmd::STATUS_SET);
    setDataLength(0x3);
}

unsigned char Radar::StatusSetCmd::getChecksum() {
    unsigned char validChecksum = Header::getChecksum(getDataLength()) + Header::getChecksum(getCmd()) +  Header::getChecksum(mode);
    checksum = validChecksum;
    return checksum;
}

void Radar::StatusSetCmd::change2SetMode()
{
    this->mode = MODE_SET;
    getChecksum();
}

void Radar::StatusSetCmd::change2SpeedTestMode()
{
    this->mode = MODE_SPEED_TEST;
    getChecksum();
}
bool Radar::StatusSetCmd::isSetMode() const
{
    return this->mode == MODE_SET;
}

bool Radar::StatusSetCmd::isSpeedTestMode() const
{
    return this->mode == MODE_SPEED_TEST;
}



Radar::MeasureSetCmd::MeasureSetCmd()
        : mode(0)
        , checksum(0)
{
    setCmd(RadarCmd::MEASUREMENT_MODE);
    setDataLength(0x3);
}

unsigned char Radar::MeasureSetCmd::getChecksum() {
    unsigned char validChecksum = Header::getChecksum(getDataLength()) + Header::getChecksum(getCmd()) +  Header::getChecksum(mode);
    checksum = validChecksum;
    return checksum;
}

void Radar::MeasureSetCmd::change2MeterMode()
{
    this->mode = METER_MODE;
    getChecksum();
}

void Radar::MeasureSetCmd::change2ExitMode()
{
    this->mode = EXIT_MODE;
    getChecksum();
}
bool Radar::MeasureSetCmd::isMeterMode() const
{
    return this->mode == METER_MODE;
}

bool Radar::MeasureSetCmd::isExitMode() const
{
    return this->mode == EXIT_MODE;
}


Radar::TestMeasureSetCmd::TestMeasureSetCmd()
        : mode(0)
        , checksum(0)
{
    setCmd(RadarCmd::TEST_MODE);
    setDataLength(0x3);
}

unsigned char Radar::TestMeasureSetCmd::getChecksum() {
    unsigned char validChecksum = Header::getChecksum(getDataLength()) + Header::getChecksum(getCmd()) +  Header::getChecksum(mode);
    checksum = validChecksum;
    return checksum;
}

void Radar::TestMeasureSetCmd::change2DirecMode()
{
    this->mode = DIREC_MODE;
    getChecksum();
}

void Radar::TestMeasureSetCmd::change2SimulateMode()
{
    this->mode = SIMULATE_MODE;
    getChecksum();
}
void Radar::TestMeasureSetCmd::change2MeasureMode()
{
    this->mode = MEATURE_MODE;
    getChecksum();
}
bool Radar::TestMeasureSetCmd::isDirecMode() const
{
    return this->mode == DIREC_MODE;
}

bool Radar::TestMeasureSetCmd::isSimulateMode() const
{
    return this->mode == SIMULATE_MODE;
}
bool Radar::TestMeasureSetCmd::isMeasureMode() const
{
    return this->mode == MEATURE_MODE;
}



Radar::SoftwareVersionSearchCmd::SoftwareVersionSearchCmd()
{
    setCmd(RadarCmd::SOFTWARE_VERSION_SEARCH);
    setDataLength(0x2);
}

Radar::SoftwareVersionSearchResponse::SoftwareVersionSearchResponse()
        : major(0)
        , minor(0)
        , revision(0)
        , checksum(0)
{}

char Radar::SoftwareVersionSearchResponse::getChecksum() const {
    return checksum;
}
const char* Radar::SoftwareVersionSearchResponse::getVersion(char* version, const int length) const
{
    if(version && length >= 12)
    {
        snprintf(version, length,"%d.%d.%d", major, minor, revision);
    }
    return version;
}

const char* Radar::SoftwareVersionSearchResponse::getVersion(std::string& version) const
{
    version.reserve(12);
    sprintf((char*)version.c_str(),"%d.%d.%d", major, minor, revision);
    return version.c_str();
}

Radar::ProductNoSearchCmd::ProductNoSearchCmd()
{
    setCmd(RadarCmd::PRODUCT_NO_SEARCH);
    setDataLength(0x2);
}


Radar::ProductNoSearchResponse::ProductNoSearchResponse() {}

const char *Radar::ProductNoSearchResponse::getProductNo() const {
    return productNo;
}

const char *Radar::ProductNoSearchResponse::getMainBoardNo() const {
    return mainBoardNo;
}

char Radar::ProductNoSearchResponse::getChecksum() {
    return checksum;
}

Radar::Parameter::Parameter()
        : radarHeight(0)
        , maxSpeed(0)
        , threshold(0)
        , hardThreshold(0)
        , triggerMode(0)
        , leftSpeed(0)
        , angle(0)
        , distance(0)
        , laneWidth(0)
        , interval(0)
        , adjustedLaneCoefficient(0)
        , roadMode(0)
        , checksum(0)
{

}

unsigned char Radar::Parameter::getRadarHeight() const {
    return radarHeight;
}

void Radar::Parameter::setRadarHeight(unsigned char radarHeight) {
    Parameter::radarHeight = radarHeight;
}

unsigned char Radar::Parameter::getMaxSpeed() const {
    return maxSpeed;
}

void Radar::Parameter::setMaxSpeed(unsigned char maxSpeed) {
    Parameter::maxSpeed = maxSpeed;
}


unsigned char Radar::Parameter::getThreshold() const {
    return threshold;
}

void Radar::Parameter::setThreshold(unsigned char threshold) {
    Parameter::threshold = threshold;
}


unsigned char Radar::Parameter::getHardThreshold() const {
    return hardThreshold;
}

void Radar::Parameter::setHardThreshold(unsigned char hardThreshold) {
    Parameter::hardThreshold = hardThreshold;
}


unsigned char Radar::Parameter::getLeftSpeed() const {
    return leftSpeed;
}

void Radar::Parameter::setLeftSpeed(unsigned char leftSpeed) {
    Parameter::leftSpeed = leftSpeed;
}

unsigned char Radar::Parameter::getAngle() const {
    return angle;
}

void Radar::Parameter::setAngle(unsigned char angle) {
    Parameter::angle = angle;
}

unsigned char Radar::Parameter::getDistance() const {
    return distance;
}

void Radar::Parameter::setDistance(unsigned char distance) {
    Parameter::distance = distance;
}

unsigned char Radar::Parameter::getLaneWidth() const {
    return laneWidth;
}

void Radar::Parameter::setLaneWidth(unsigned char laneWidth) {
    Parameter::laneWidth = laneWidth;
}


unsigned char Radar::Parameter::getInterval() const {
    return interval;
}

void Radar::Parameter::setInterval(unsigned char interval) {
    Parameter::interval = interval;
}


unsigned char Radar::Parameter::getAdjustedLaneCoefficient() const {
    return adjustedLaneCoefficient;
}

void Radar::Parameter::setAdjustedLaneCoefficient(unsigned char adjustedLaneCoefficient) {
    Parameter::adjustedLaneCoefficient = adjustedLaneCoefficient;
}


unsigned char Radar::Parameter::getChecksum() const {
    return checksum;
}

Radar::Parameter::TriggerMode Radar::Parameter::getTriggerMode() const {
    return (Radar::Parameter::TriggerMode)triggerMode;
}

void Radar::Parameter::setTriggerMode(TriggerMode triggerMode) {
    Parameter::triggerMode = triggerMode;
}

void Radar::Parameter::setTriggerMode(unsigned char triggerMode)
{
    if( triggerMode >=TRIGGER_HEADER && triggerMode <= TRIGGER_CONTINUE )
        Parameter::triggerMode = (Radar::Parameter::TriggerMode)triggerMode;
    else
        Parameter::triggerMode = TRIGGER_HEADER;
}



Radar::Parameter::RoadMode Radar::Parameter::getRoadMode() const {
    return (Radar::Parameter::RoadMode)roadMode;
}

void Radar::Parameter::setRoadMode(RoadMode roadMode) {
    Parameter::roadMode = roadMode;
}

void Radar::Parameter::setRoadMode(unsigned char roadMode)
{
    if( roadMode >= ROAD_CITY && roadMode <= ROAD_HIGHWAY )
        Parameter::roadMode = (Radar::Parameter::RoadMode)roadMode;
    else
        Parameter::roadMode = ROAD_CITY;
}


int Radar::Parameter::isValidChecksum() const
{
    unsigned char validChecksum = getChecksum();

    return checksum == (unsigned char)validChecksum ? 0 : ERROR_CHECKSUM;
}

unsigned char Radar::Parameter::getChecksum()
{
    short validChecksum = Header::getChecksum(getCmd())
                          + getDataLength()
                          + radarHeight
                          + maxSpeed
                          + threshold
                          + hardThreshold
                          + triggerMode
                          + leftSpeed
                          + angle
                          + distance // 雷达到车道的距离
                          + laneWidth
                          + interval   //两车间的间隙
                          + adjustedLaneCoefficient
                          + roadMode;

    return (checksum = (unsigned char)validChecksum);
}

int Radar::Parameter::isValidStatus() const
{
    return 0;
}
bool Radar::Parameter::isSuccess() const
{
    return 0 == isValidStatus();
}

bool Radar::Parameter::isValidTriggerMode(int mode)
{
    return mode >= TRIGGER_HEADER && mode <= TRIGGER_CONTINUE;
}

bool Radar::Parameter::isValidRoadMode(int mode)
{
    return ROAD_CITY   <= mode && ROAD_HIGHWAY >= mode;
}
Radar::ParameterSaveCmd::ParameterSaveCmd()
{
    setCmd(RadarCmd::PRAMETER_SAVE);
    setDataLength(0x2);
}

Radar::ParameterGetCmd::ParameterGetCmd()
{
    setCmd(RadarCmd::PRAMETER_GET);
    setDataLength(0x2);
    getChecksum();
}

Radar::ParameterSetCmd::ParameterSetCmd()
{
    setCmd(RadarCmd::PRAMETER_SET);
    setDataLength(0x0E);

}

void Radar::ParameterSetCmd::setParameter(const Radar::Parameter* parameter)
{
    if(parameter)
    {
        this->setTriggerMode( parameter->getTriggerMode() );
        this->setRadarHeight(parameter->getRadarHeight());
        this->setMaxSpeed(parameter->getMaxSpeed());
        this->setThreshold(parameter->getThreshold());
        this->setHardThreshold(parameter->getHardThreshold());
        this->setLeftSpeed(parameter->getLeftSpeed());
        this->setAngle(parameter->getAngle());
        this->setDistance(parameter->getDistance());
        this->setLaneWidth(parameter->getLaneWidth());
        this->setInterval(parameter->getInterval());
        this->setAdjustedLaneCoefficient(this->getAdjustedLaneCoefficient());
        this->setRoadMode(parameter->getRoadMode());
    }
}

Radar::SpeedResponse::SpeedResponse()
    : speed(0)
    , lane(0)
    , direction(0)
    , checksum(0)
{}

unsigned short Radar::SpeedResponse::getSpeed() const {
    return speed;
}

unsigned char Radar::SpeedResponse::getLane() const {
    return lane;
}

unsigned char Radar::SpeedResponse::getDirection() const {
    return direction;
}

unsigned char Radar::SpeedResponse::getChecksum() const {
    return checksum;
}

bool Radar::SpeedResponse::isValidChecksum() const
{
    unsigned char validChecksum = Header::getChecksum(getCmd())
                                    + getDataLength()
                                    + Header::getChecksum(speed)
                                    + lane
                                    + direction;

    return checksum == validChecksum;
}

File: BchtSystemSpeed/app/src/main/java/com/huang/bchtsystem/rader/RaderCmd.java
package com.huang.bchtsystem.rader;

/**
 * Created by admin on 2017/6/7.
 */

public class RaderCmd {

    final public static short Null = (short) 0x8310;
    final public static short STATUS_SEARCH = (short) 0x8310;
    final public static short RESET = (short) 0x8320;
    final public static short STATUS_SET = (short) 0x8350;
    final public static short SOFTWARE_VERSION_SEARCH = (short) 0x8351;
    final public static short PRODUCT_NO_SEARCH = (short) 0x8352;
    final public static short PRAMETER_SET = (short) 0x8355;
    final public static short PRAMETER_GET = (short) 0x8356;
    final public static short PRAMETER_SAVE = (short) 0x8357;
    final public static short RESTORE = (short) 0x8359;

    final public static short STATUS_RESPONSE_SUCCESS = (short) 0x853A;
    final public static short STATUS_RESPONSE_FAILURE = (short) 0x85FF;

    final public static byte HEADER1 = (byte) 0xFF;
    final public static byte HEADER2 = (byte) 0xFA;

    final public static byte POSITIVE = (byte) 0x00;
    final public static byte NEGATIVE = (byte) 0x01;

    final public static int ERROR_BASE = 0x1000;
    final public static int ERROR_UNKNOWN = (ERROR_BASE + 0x1);
    final public static int ERROR_HEADER1 = (ERROR_BASE + 0x2);
    final public static int ERROR_HEADER2 = (ERROR_BASE + 0x3);
    final public static int ERROR_DATALENGTH = (ERROR_BASE + 0x4);
    final public static int ERROR_CHECKSUM = (ERROR_BASE + 0x5);
    final public static int ERROR_RESPONSE_STATUS = (ERROR_BASE + 0x6);
    final public static int ERROR_CMD = (ERROR_BASE + 0x7);
    final public static int ERROR_FAILURE = (ERROR_BASE + 0x8);
    final public static int ERROR_NULL = (ERROR_BASE + 0x9);
    final public static int ERROR_NOT_OPENED = (ERROR_BASE + 0xA);
    final public static int ERROR_SPEED = (ERROR_BASE + 0xB);
    final public static int ERROR_CONFIG = (ERROR_BASE + 0xC);
    final public static int ERROR_PARAM_NOT_SAVED = (ERROR_BASE + 0xD);
    final public static int ERROR_WORKING_MODE = (ERROR_BASE + 0xE);
    final public static int ERROR_ANGLE = (ERROR_BASE + 0x10);
    final public static int ERROR_PARAM_OUT_OF_BOUND = (ERROR_BASE + 0x11);

    public class WorkingMode {
        final public static byte MODE_SET = 0x01;
        final public static byte MODE_SPEED_TEST = 0x02;
    };

    public class Mode {
        final public static byte METER_MODE = 0x01;
        final public static byte EXIT_MODE = 0x02;
    };

    public enum TriggerMode {
        TRIGGER_HEADER,
        TRIGGER_TAIL,
        TRIGGER_BOTH,
        TRIGGER_CONTINUE
    };

    public enum RoadMode {
        ROAD_CITY,
        ROAD_HIGHWAY,
    };

    public class OutOfBoundException extends Exception
    {
        public OutOfBoundException()
        {
            super();
        }

        public OutOfBoundException(String message)
        {
            super(message);
        }
    };

    public class Response
    {

        public Response()
        {
            status = 0;
        }

        public short getStatus()
        {
            return status;
        }

        public int isValidStatus()
        {
            return (this.status == RaderCmd.STATUS_RESPONSE_SUCCESS || this.status == RaderCmd.STATUS_RESPONSE_FAILURE ) ? 0 : ERROR_RESPONSE_STATUS;
        }
        public boolean isSuccess()
        {
            return this.status == RaderCmd.STATUS_RESPONSE_SUCCESS;
        }

        private short   status;
    };

    public class SpeedResponse
    {
        public short getSpeed() {
            return speed;
        }

        public void setSpeed(short speed) {
            this.speed = speed;
        }

        public char getLane() {
            return lane;
        }

        public void setLane(char lane) {
            this.lane = lane;
        }

        public char getDirection() {
            return direction;
        }

        public void setDirection(char direction) {
            this.direction = direction;
        }

        private short speed;
        private char lane;
        private char direction;
        //private char checksum;
    }

    class StatusSetCmd
    {

        public StatusSetCmd()
        {
            mode = 0;
        }

        public void change2SetMode()
        {
            this.mode = WorkingMode.MODE_SET;
        }
        public void change2SpeedTestMode()
        {
            this.mode = WorkingMode.MODE_SPEED_TEST;
        }
        public boolean isSetMode()
        {
            return this.mode == WorkingMode.MODE_SET;
        }
        public boolean isSpeedTestMode()
        {
            return this.mode == WorkingMode.MODE_SPEED_TEST;
        }

        private byte    mode;

    };

    public class StatusSetResponse extends Response
     {

     };


    class MeasureSetCmd
    {

        public MeasureSetCmd()
        {
            mode = 0;
        }

        public void change2MeterMode()
        {
            this.mode = Mode.METER_MODE;
        }
        public void change2ExitMode()
        {
            this.mode = Mode.EXIT_MODE;
        }
        public boolean isMeterMode()
        {
            return this.mode == Mode.METER_MODE;
        }
        public boolean isExitMode()
        {
            return this.mode == Mode.EXIT_MODE;
        }

        private byte    mode;

    };

    public class MeasureSetResponse extends Response
    {

    };


    class ParameterGetCmd
    {

        public ParameterGetCmd()
        {
        }
    };

    public  class Parameter
    {

        public Parameter()
        {
            raderHeight=0;
            maxSpeed=0;
            threshold=0;
            hardThreshold=0;
            triggerMode=0;
            leftSpeed=0;
            angle=0;
            distance=0;
            laneWidth=0;
            interval=0;
            adjustedLaneCoefficient=0;
            roadMode=0;
        }

        public float getRaderHeight() {
            return raderHeight;
        }

        public void setRaderHeight(float raderHeight) throws Exception{
            if( raderHeight >= 0 && raderHeight <= 25.5)
                this.raderHeight = raderHeight;
            else
                throw new Exception("ERROR_PARAM_OUT_OF_BOUND");
        }

        public float getMaxSpeed() {
            return maxSpeed;
        }

        public void setMaxSpeed(float maxSpeed) {
            this.maxSpeed = maxSpeed;
        }

        public float getThreshold() {
            return threshold;
        }

        public void setThreshold(float threshold) {
            this.threshold = threshold;
        }

        public float getHardThreshold() {
            return hardThreshold;
        }

        public void setHardThreshold(float hardThreshold) throws Exception{

            if( hardThreshold >= 0 && hardThreshold <= 25.5)
                this.hardThreshold = hardThreshold;
            else
                throw new Exception("ERROR_PARAM_OUT_OF_BOUND");
        }

        public float getTriggerMode() {
            return triggerMode;
        }

        public void setTriggerMode(TriggerMode triggerMode) {
            switch (triggerMode) {
                case TRIGGER_HEADER:
                    this.triggerMode = 0x00;
                    break;
                case TRIGGER_TAIL:
                    this.triggerMode = 0x01;
                    break;
                case TRIGGER_BOTH:
                    this.triggerMode = 0x02;
                    break;
                case TRIGGER_CONTINUE:
                    this.triggerMode = 0x03;
                    break;
                default:
                    this.triggerMode = 0x00;
                    break;
            }
        }

        public float getLeftSpeed() {
            return leftSpeed;
        }

        public void setLeftSpeed(float leftSpeed) {
            this.leftSpeed = leftSpeed;
        }

        public float getAngle() {
            return angle;
        }

        public void setAngle(float angle) {
            this.angle = angle;
        }

        public float getDistance() {
            return distance;
        }

        public void setDistance(float distance) throws Exception {

            if( distance >= 0 && distance <= 25.5 )
                this.distance = distance;
            else
                throw new Exception("ERROR_PARAM_OUT_OF_BOUND");
        }

        public float getLaneWidth() {
            return laneWidth;
        }

        public void setLaneWidth(float laneWidth) throws Exception {
            if( laneWidth >= 0 && laneWidth <= 25.5 )
                this.laneWidth = laneWidth;
            else
                throw new Exception("ERROR_PARAM_OUT_OF_BOUND");
        }

        public float getInterval() {
            return interval;
        }

        public void setInterval(float interval) throws Exception {
            if( interval >=0 && interval <= 25.5 )
                this.interval = interval;
            else
                throw new Exception("ERROR_PARAM_OUT_OF_BOUND");
        }

        public float getAdjustedLaneCoefficient() {
            return adjustedLaneCoefficient;
        }

        public void setAdjustedLaneCoefficient(float adjustedLaneCoefficient) throws Exception {
            if( adjustedLaneCoefficient >= 0 && adjustedLaneCoefficient <= 25.5 )
                this.adjustedLaneCoefficient = adjustedLaneCoefficient;
            else
                throw new Exception("ERROR_PARAM_OUT_OF_BOUND");
        }

        public float getRoadMode() {
            return roadMode;
        }

        public void setRoadMode(RoadMode roadMode) {
            switch (roadMode)
            {
                case ROAD_CITY:
                    this.roadMode = 0x00;
                    break;
                case ROAD_HIGHWAY:
                    this.roadMode = 0x01;
                    break;
                default:
                    this.roadMode = 0x00;
                    break;
            }
        }

        private float    raderHeight;
        private float    maxSpeed;
        private float    threshold;
        private float    hardThreshold;
        private float    triggerMode;
        private float    leftSpeed;
        private float    angle;
        private float    distance; // 雷达到车道的距离
        private float    laneWidth;
        private float    interval;   //两车间的间隙
        private float    adjustedLaneCoefficient;
        private float    roadMode;
    };

    public class ParameterGetResponse
    {
        public Parameter parameter;

        public Parameter getParameter() {
            return parameter;
        }

        public void setParameter(Parameter parameter) {
            this.parameter = parameter;
        }
    };

    public class ParameterSetCmd
    {
        public Parameter parameter;

        public ParameterSetCmd()
        {

        }

        public Parameter getParameter() {
            return parameter;
        }

        public void setParameter(Parameter parameter) {
            this.parameter = parameter;
        }
    };

    class ParameterSetResponse extends Response
    {

    };
}
